curriculum/locales/english/learn-digital-signatures-by-building-a-wallet.md

```diff
@@ -4,206 +4,5928 @@
+
+#### --"blockchain-helpers.js"--
+
+```js
+import sha256 from 'crypto-js/sha256.js';
+import EC from 'elliptic';
+import { writeFileSync, readFileSync } from 'fs';
+
+const ec = new EC.ec('p192');
+
+export function writeBlockchain(blockchain) {
+  const blockchainString = JSON.stringify(blockchain, null, 2);
+  writeFileSync('./blockchain.json', blockchainString);
+}
+
+export function getBlockchain() {
+  const blockchainFile = readFileSync('./blockchain.json');
+  const blockchain = JSON.parse(blockchainFile);
+  return blockchain;
+}
+
+export function isValidChain() {
+  const blockchain = getBlockchain();
+
+  // loop through blocks
+  for (let i = 1; i < blockchain.length; i++) {
+    const previousBlock = blockchain[i - 1];
+    const { hash, nonce, previousHash, transactions } = blockchain[i];
+
+    // validate previous hash
+    if (previousHash !== previousBlock.hash) {
+      return false;
+    }
+
+    // validate block hash
+    const testBlockHash = sha256(nonce + previousBlock.hash + JSON.stringify(transactions)).toString();
+    if (hash != testBlockHash) {
+      return false;
+    }
+
+    // loop through transactions
+    for (let j = 0; j < transactions.length; j++) {
+      const { hash, fromAddress, toAddress, amount, signature } = transactions[j];
+
+      // don't validate reward transactions
+      if (fromAddress !== null) {
+
+        // validate transaction hash
+        const testTransactionHash = sha256(fromAddress + toAddress + amount).toString();
+        if (hash != testTransactionHash) {
+          return false;
+        }
+
+        // validate signature
+        const publicKeyPair = ec.keyFromPublic(fromAddress, 'hex');
+
+      }
+    }
+  }
+
+  return true;
+}
+
+export function writeTransactions(transactions) {
+  const transactionsString = JSON.stringify(transactions, null, 2);
+  writeFileSync('./transactions.json', transactionsString);
+}
+
+export function getTransactions() {
+  const transactionsFile = readFileSync('./transactions.json');
+  const transactions = JSON.parse(transactionsFile);
+  return transactions;
+}
+
+export function getAddressBalance(address) {
+  const blockchain = getBlockchain();
+  let balance = 0;
+
+  // loop through blocks
+  for (let i = 1; i < blockchain.length; i++) { 
+    const { transactions } = blockchain[i];
+
+    // loop through transactions
+    for (let j = 0; j < transactions.length; j++) {
+      const { fromAddress, toAddress, amount } = transactions[j];
+
+      if (fromAddress === address) {
+        balance -= amount;
+      }
+
+      if (toAddress === address) {
+        balance += amount;
+      }
+    }
+  }
+
+  return balance;
+}
+
+export function getWalletAddressFromName(name) {
+  const walletsFile = readFileSync('./wallets.json');
+  const wallets = JSON.parse(walletsFile);
+  return wallets[name];
+}
+```
+
+## 108
+
+### --description--
+
+Add an `if` condition that checks if `!verifiedSignature`. If it's not valid, return `false`.
```

Comment:
Just for less copy-pasta 🤷‍♂️ 

```suggestion
Add an `if` condition that checks if the signature is **not** verified. If it's not valid, return `false`.
```

Reply:
I originally worded it close to this, but the test checks for that specific condition so I changed it cause there's too many other ways to make that same check. I will see if I can improve it a little.